File: bindings/dotnet/AgentStateGraph/Repository.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using AgentStateGraph.Interop;

namespace AgentStateGraph;

/// <summary>
/// An AgentStateGraph repository — an AI-native, commit-versioned state
/// store. The handle is <see cref="IDisposable"/>; always dispose (ideally
/// via <c>using</c>) to release the underlying native resources promptly.
/// </summary>
/// <remarks>
/// Construct with <see cref="Repository()"/> for an ephemeral in-memory
/// store or <see cref="OpenSqlite(string)"/> for a SQLite-backed durable
/// store. Derived handles (<see cref="TaskStore"/>, <see cref="PolicyStore"/>)
/// hold a reference back to this instance to keep the repository alive
/// for the lifetime of any derived store.
/// </remarks>
public sealed class Repository : IDisposable
{
    private readonly SafeRepoHandle _handle;
    private bool _disposed;

    /// <summary>Default ref for all operations when a caller omits one.</summary>
    public const string DefaultRef = "main";

    /// <summary>Creates an ephemeral, in-memory repository.</summary>
    public Repository()
    {
        var raw = NativeMethods.agentstategraph_new_memory();
        if (raw == IntPtr.Zero)
        {
            throw new AgentStateGraphException("new", "failed to create memory repository");
        }
        _handle = SafeRepoHandle.Adopt(raw);
    }

    private Repository(SafeRepoHandle handle)
    {
        _handle = handle;
    }

    /// <summary>Opens (or creates) a SQLite-backed repository at <paramref name="path"/>.</summary>
    public static Repository OpenSqlite(string path)
    {
        ArgumentNullException.ThrowIfNull(path);
        var raw = NativeMethods.agentstategraph_new_sqlite(path);
        if (raw == IntPtr.Zero)
        {
            throw new AgentStateGraphException("open_sqlite", $"failed to open sqlite repository at '{path}'");
        }
        return new Repository(SafeRepoHandle.Adopt(raw));
    }

    /// <summary>Internal accessor used by <see cref="TaskStore"/> / <see cref="PolicyStore"/>.</summary>
    internal SafeRepoHandle Handle => _handle;

    private void ThrowIfDisposed()
    {
        if (_disposed)
        {
            throw new ObjectDisposedException(nameof(Repository));
        }
    }

    // -----------------------------------------------------------------------
    // Read / write operations
    // -----------------------------------------------------------------------

    /// <summary>Returns the JSON value at <paramref name="path"/> as a string.</summary>
    public string Get(string path, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_get(_handle.DangerousGetHandle(), refName, path);
        var raw = Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("get", "native FFI returned null");
        return raw;
    }

    /// <summary>
    /// Writes <paramref name="jsonValue"/> at <paramref name="path"/> and
    /// creates a new commit. Returns the short commit id.
    /// </summary>
    public CommitId Set(
        string path,
        string jsonValue,
        string intentCategory,
        string intentDescription,
        string refName = DefaultRef)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_set(
            _handle.DangerousGetHandle(), refName, path, jsonValue, intentCategory, intentDescription);
        var raw = Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("set", "native FFI returned null");
        return new CommitId(raw);
    }

    /// <summary>
    /// Convenience: serializes <paramref name="value"/> to JSON via the
    /// shared <see cref="JsonSerializer"/> options, then delegates to
    /// <see cref="Set(string, string, string, string, string)"/>.
    /// </summary>
    public CommitId SetJson<T>(
        string path,
        T value,
        string intentCategory,
        string intentDescription,
        string refName = DefaultRef)
    {
        var json = Json.Serialize(value);
        return Set(path, json, intentCategory, intentDescription, refName);
    }

    /// <summary>Removes the value at <paramref name="path"/> and creates a new commit.</summary>
    public CommitId Delete(
        string path,
        string intentCategory,
        string intentDescription,
        string refName = DefaultRef)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_delete(
            _handle.DangerousGetHandle(), refName, path, intentCategory, intentDescription);
        var raw = Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("delete", "native FFI returned null");
        return new CommitId(raw);
    }

    /// <summary>
    /// Creates a new branch <paramref name="name"/> starting at
    /// <paramref name="from"/> (or the default ref if null).
    /// </summary>
    public string Branch(string name, string? from = null)
    {
        ThrowIfDisposed();
        // FFI expects a non-null `from`; substitute the default ref when
        // the caller omits it. (The Go binding passes "main" explicitly;
        // the C# API models the default as an optional parameter.)
        var fromRef = from ?? "main";
        var ptr = NativeMethods.agentstategraph_branch(_handle.DangerousGetHandle(), name, fromRef);
        return Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("branch", "native FFI returned null");
    }

    /// <summary>
    /// Returns every branch whose name starts with <paramref name="prefix"/>.
    /// Pass <c>null</c> (or an empty string) for no filter.
    /// </summary>
    public IReadOnlyList<BranchEntry> ListBranches(string? prefix = null)
    {
        ThrowIfDisposed();
        var filter = string.IsNullOrEmpty(prefix) ? null : prefix;
        var ptr = NativeMethods.agentstategraph_list_branches(_handle.DangerousGetHandle(), filter);
        var raw = Strings.ConsumeUtf8(ptr);
        return Json.Deserialize<List<BranchEntry>>(raw, "list_branches");
    }

    /// <summary>
    /// Removes the branch <paramref name="name"/>. Returns <c>true</c> if
    /// the branch existed, <c>false</c> if it did not (non-error).
    /// </summary>
    public bool DeleteBranch(string name)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(name);
        var ptr = NativeMethods.agentstategraph_delete_branch(_handle.DangerousGetHandle(), name);
        var raw = Json.ThrowIfError(Strings.ConsumeUtf8(ptr), "delete_branch");
        try
        {
            using var doc = JsonDocument.Parse(raw);
            if (doc.RootElement.ValueKind == JsonValueKind.Object
                && doc.RootElement.TryGetProperty("deleted", out var deleted)
                && deleted.ValueKind is JsonValueKind.True or JsonValueKind.False)
            {
                return deleted.GetBoolean();
            }
            throw new AgentStateGraphException("delete_branch", $"unexpected response: {raw}");
        }
        catch (JsonException ex)
        {
            throw new AgentStateGraphException("delete_branch", $"failed to parse response: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Computes a structured diff between two refs. Returns raw JSON; the
    /// shape mirrors the Rust <c>Diff</c> type and is left undecoded here
    /// to match the Go binding's deliberate pass-through.
    /// </summary>
    public string Diff(string refA, string refB)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_diff(_handle.DangerousGetHandle(), refA, refB);
        return Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("diff", "native FFI returned null");
    }

    /// <summary>
    /// Merges <paramref name="source"/> into <paramref name="target"/>.
    /// Returns the short commit id of the merge commit.
    /// </summary>
    public CommitId Merge(string source, string target, string description)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_merge(
            _handle.DangerousGetHandle(), source, target, description);
        var raw = Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("merge", "native FFI returned null");
        // The FFI surfaces merge errors as "error:<msg>" prefixed strings
        // (agentstategraph.go does the same defensive check).
        if (raw.StartsWith("error:", StringComparison.Ordinal))
        {
            throw new AgentStateGraphException("merge", raw[6..]);
        }
        return new CommitId(raw);
    }

    /// <summary>
    /// Returns up to <paramref name="limit"/> commits on
    /// <paramref name="refName"/>, newest first.
    /// </summary>
    public IReadOnlyList<Commit> Log(uint limit = 100, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_log(_handle.DangerousGetHandle(), refName, limit);
        var raw = Strings.ConsumeUtf8(ptr);
        return Json.Deserialize<List<Commit>>(raw, "log");
    }

    /// <summary>
    /// Returns who last modified <paramref name="path"/> and why. Raw JSON
    /// blame entry — shape is defined by <c>agentstategraph_core::Blame</c>.
    /// </summary>
    public string Blame(string path, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        var ptr = NativeMethods.agentstategraph_blame(_handle.DangerousGetHandle(), refName, path);
        return Strings.ConsumeUtf8(ptr)
            ?? throw new AgentStateGraphException("blame", "native FFI returned null");
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }
        _disposed = true;
        _handle.Dispose();
    }
}

File: bindings/swift/Sources/AgentStateGraph/Repository.swift
import Foundation
import CAgentStateGraph

/// Intent categories understood by the AgentStateGraph commit layer.
///
/// Every mutation records *why* it happened. Pass one of these to `set`,
/// `delete`, etc. Unknown strings fall back to the native default, so you
/// may also use `.other("myCategory")` for project-specific intents.
public enum IntentCategory: RawRepresentable, Sendable, Equatable {
    case checkpoint
    case correction
    case refinement
    case exploration
    case merge
    case rollback
    case other(String)

    public init(rawValue: String) {
        switch rawValue.lowercased() {
        case "checkpoint": self = .checkpoint
        case "correction": self = .correction
        case "refinement": self = .refinement
        case "exploration": self = .exploration
        case "merge": self = .merge
        case "rollback": self = .rollback
        default: self = .other(rawValue)
        }
    }

    public var rawValue: String {
        switch self {
        case .checkpoint: return "Checkpoint"
        case .correction: return "Correction"
        case .refinement: return "Refinement"
        case .exploration: return "Exploration"
        case .merge: return "Merge"
        case .rollback: return "Rollback"
        case .other(let s): return s
        }
    }
}

/// One row returned by `listBranches`.
public struct BranchEntry: Codable, Sendable, Equatable {
    public let name: String
    public let target: String
}

/// One row returned by `log`.
public struct LogEntry: Codable, Sendable, Equatable {
    public let id: String
    public let agent: String
    public let intentCategory: String
    public let intentDescription: String
    public let reasoning: String?
    public let confidence: Double?

    enum CodingKeys: String, CodingKey {
        case id, agent, reasoning, confidence
        case intentCategory = "intent_category"
        case intentDescription = "intent_description"
    }
}

/// A handle to an AgentStateGraph repository — an AI-native, versioned,
/// intent-carrying state store.
///
/// ```swift
/// let asg = try AgentStateGraph()          // in-memory
/// try asg.set("/name", json: "\"my-cluster\"", category: .checkpoint, description: "init")
/// let name = try asg.get("/name")          // "\"my-cluster\""
/// asg.close()
/// ```
///
/// The handle owns native memory. Call `close()` when done, or rely on
/// `deinit`. Not safe to use after `close()`.
public final class AgentStateGraph {
    private var repo: UnsafeMutableRawPointer?

    init(repo: UnsafeMutableRawPointer?) throws {
        guard let repo = repo else {
            throw AgentStateGraphError.operationFailed("open repository")
        }
        self.repo = repo
    }

    /// Create a new in-memory (ephemeral) repository.
    public convenience init() throws {
        try self.init(repo: agentstategraph_new_memory())
    }

    /// Open (or create) a durable SQLite-backed repository at `path`.
    public convenience init(path: String) throws {
        let c = sgDup(path); defer { free(c) }
        try self.init(repo: agentstategraph_new_sqlite(c))
    }

    deinit { close() }

    /// Free the underlying repository. Idempotent.
    public func close() {
        if let r = repo {
            agentstategraph_free(r)
            repo = nil
        }
    }

    private func handle() throws -> UnsafeMutableRawPointer {
        guard let r = repo else { throw AgentStateGraphError.closed("repository") }
        return r
    }

    /// Raw pointer for sibling stores (TaskStore/PolicyStore). Internal.
    func rawHandle() throws -> UnsafeMutableRawPointer { try handle() }

    // MARK: - Values

    /// Return the JSON value at `path` (as a JSON string) on `ref`.
    public func get(_ path: String, ref: String = "main") throws -> String {
        let r = try handle()
        let cRef = sgDup(ref); defer { free(cRef) }
        let cPath = sgDup(path); defer { free(cPath) }
        return try consume(agentstategraph_get(r, cRef, cPath), "get")
    }

    /// Decode the JSON value at `path` into a `Decodable` type.
    public func get<T: Decodable>(_ path: String, as type: T.Type, ref: String = "main") throws -> T {
        try decodeJSON(try get(path, ref: ref))
    }

    /// Write a raw JSON string at `path`, recording intent. Returns the commit id.
    @discardableResult
    public func set(
        _ path: String,
        json: String,
        category: IntentCategory,
        description: String,
        ref: String = "main"
    ) throws -> String {
        let r = try handle()
        let cRef = sgDup(ref); defer { free(cRef) }
        let cPath = sgDup(path); defer { free(cPath) }
        let cVal = sgDup(json); defer { free(cVal) }
        let cCat = sgDup(category.rawValue); defer { free(cCat) }
        let cDesc = sgDup(description); defer { free(cDesc) }
        return try consume(agentstategraph_set(r, cRef, cPath, cVal, cCat, cDesc), "set")
    }

    /// Encode `value` to JSON and write it at `path`. Returns the commit id.
    @discardableResult
    public func set<T: Encodable>(
        _ path: String,
        value: T,
        category: IntentCategory,
        description: String,
        ref: String = "main"
    ) throws -> String {
        try set(path, json: try encodeJSON(value), category: category, description: description, ref: ref)
    }

    /// Delete the value at `path`, recording intent. Returns the commit id.
    @discardableResult
    public func delete(
        _ path: String,
        category: IntentCategory,
        description: String,
        ref: String = "main"
    ) throws -> String {
        let r = try handle()
        let cRef = sgDup(ref); defer { free(cRef) }
        let cPath = sgDup(path); defer { free(cPath) }
        let cCat = sgDup(category.rawValue); defer { free(cCat) }
        let cDesc = sgDup(description); defer { free(cDesc) }
        return try consume(agentstategraph_delete(r, cRef, cPath, cCat, cDesc), "delete")
    }

    // MARK: - Branches

    /// Create branch `name` starting from ref `from`. Returns the commit id.
    @discardableResult
    public func branch(_ name: String, from: String) throws -> String {
        let r = try handle()
        let cName = sgDup(name); defer { free(cName) }
        let cFrom = sgDup(from); defer { free(cFrom) }
        return try consume(agentstategraph_branch(r, cName, cFrom), "branch")
    }

    /// List branches whose name starts with `prefix` (pass `nil` for all).
    public func listBranches(prefix: String? = nil) throws -> [BranchEntry] {
        let r = try handle()
        let cPrefix = sgDup(prefix); defer { free(cPrefix) }
        let raw = try consume(agentstategraph_list_branches(r, cPrefix), "list_branches")
        return try decodeJSON(raw)
    }

    /// Delete branch `name`. Returns `true` if it existed, `false` if not.
    @discardableResult
    public func deleteBranch(_ name: String) throws -> Bool {
        let r = try handle()
        let cName = sgDup(name); defer { free(cName) }
        let raw = try consume(agentstategraph_delete_branch(r, cName), "delete_branch")
        struct Resp: Decodable { let deleted: Bool }
        return try decodeJSON(raw, as: Resp.self).deleted
    }

    // MARK: - History / inspection

    /// Structured diff between two refs, as a JSON string.
    public func diff(_ refA: String, _ refB: String) throws -> String {
        let r = try handle()
        let cA = sgDup(refA); defer { free(cA) }
        let cB = sgDup(refB); defer { free(cB) }
        return try consume(agentstategraph_diff(r, cA, cB), "diff")
    }

    /// Merge `source` into `target`. Returns the merge commit id.
    @discardableResult
    public func merge(source: String, target: String, description: String) throws -> String {
        let r = try handle()
        let cSrc = sgDup(source); defer { free(cSrc) }
        let cTgt = sgDup(target); defer { free(cTgt) }
        let cDesc = sgDup(description); defer { free(cDesc) }
        return try consume(agentstategraph_merge(r, cSrc, cTgt, cDesc), "merge")
    }

    /// Commit log for `ref`, most recent first, up to `limit` entries.
    public func log(limit: UInt32 = 50, ref: String = "main") throws -> [LogEntry] {
        let r = try handle()
        let cRef = sgDup(ref); defer { free(cRef) }
        let raw = try consume(agentstategraph_log(r, cRef, limit), "log")
        return try decodeJSON(raw)
    }

    /// Who last modified `path`, and why — as a JSON string.
    public func blame(_ path: String, ref: String = "main") throws -> String {
        let r = try handle()
        let cRef = sgDup(ref); defer { free(cRef) }
        let cPath = sgDup(path); defer { free(cPath) }
        return try consume(agentstategraph_blame(r, cRef, cPath), "blame")
    }
}
